### DISCUSSION TITLE
Using `parse_string` from examples

### DISCUSSION THREAD

**Discussion by sukhmel** [winnow-rs/winnow#272] [Q&A] [answered]
I found myself needing a parser that would allow for parsing a quoted string, and sure enough it is present in the examples within the Elements of Programming Languages topic.

What is unclear to me is how that example is supposed to be used. Should everyone in such a situation copy the parser over to their codebase? Why is it only an example and not a part of the `winnow` sources? It seems like this has to do with the idea of tailoring stuff for one's own needs, but some of the things seem to be general enough to be useful on their own without the need for fine-tuning.

**Comment by epage** [comment, marked as answer]
The example is meant to show how things *can* be done.  How they are actually done is dependent on your grammar. 

More generally, we've not yet defined what our guidelines are for what parsers should be added but are mostly at "whatever we carried over from `nom`".  There are trade offs, like
- How universal does the grammar (escaping rules, quote types like python makes `'` and `"` interchangeable, etc) need to be for us to accept a parser?
- The more niche parsers we add, the slower compile times will be, the harder it will be to browse the docs, and the more work it will be to make changes to `winnow`

For myself, I find some of the built-in `winnow::ascii` parsers useful for prototyping but when I'm dealing with a more formal grammar in production, I like to have the calls to `winnow` match the BNF and so I hand write everything ([example](https://github.com/toml-rs/toml/blob/main/crates/toml_edit/src/parser/trivia.rs)).  For example `winnow::ascii::float` is meant to accept `infinity` (#265) but the [TOML grammar](https://github.com/toml-lang/toml/blob/main/toml.abnf#L165) does not.

**Comment by sukhmel** [comment]
Well, I agree with your points, I guess I've underestimated how fine may be
the necessary tuning of things and likely overestimated how common a use
case of C-style quote parsing is.

Thank you for your time explaining the reasoning and pointing out that
there still may be changes in the exposed parsers in the future.

For now, I'll stick to the example as it suits my needs, but will make sure
to check out the link you provided as I am currently trying to understand
how to map the idea I have in rough BNF-like definitions to the code 😅

**Reply by epage** [reply]
> likely overestimated how common a use
case of C-style quote parsing is.

For just c-style escapes without a specific policy, we do offer
- https://docs.rs/winnow/latest/winnow/ascii/fn.escaped.html
- https://docs.rs/winnow/latest/winnow/ascii/fn.escaped_transform.html

The examples should be updated to link to relevant code...

**Reply by epage** [reply]
Docs updated in #273
